File: agentos_cta/utils/token/budget.py
# Copyright (c) 2026 SPHARX. All Rights Reserved. "From data intelligence emerges."
# Token 预算管理。

import time
from typing import Optional, Dict, Any
from dataclasses import dataclass
from agentos_cta.utils.error import ResourceLimitError


@dataclass
class TokenUsage:
    """Token 使用情况。"""
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    timestamp: float = 0.0


class TokenBudget:
    """
    Token 预算管理器。
    跟踪任务级别的 Token 消耗，支持预算控制和预警。
    """

    def __init__(self, max_tokens: int, soft_limit: Optional[int] = None):
        """
        初始化预算管理器。

        Args:
            max_tokens: 硬性上限，超过则抛出异常。
            soft_limit: 软性上限，超过则记录警告。
        """
        self.max_tokens = max_tokens
        self.soft_limit = soft_limit or int(max_tokens * 0.8)
        self.usage = TokenUsage()
        self.history: list[TokenUsage] = []

    def add(self, input_tokens: int, output_tokens: int):
        """添加 Token 消耗。"""
        self.usage.input_tokens += input_tokens
        self.usage.output_tokens += output_tokens
        self.usage.total_tokens += input_tokens + output_tokens
        self.usage.timestamp = time.time()

        if self.usage.total_tokens > self.max_tokens:
            raise ResourceLimitError(
                f"Token budget exceeded: {self.usage.total_tokens} > {self.max_tokens}"
            )
        elif self.usage.total_tokens > self.soft_limit:
            # 记录警告，但不中断
            pass

    def checkpoint(self) -> TokenUsage:
        """创建检查点并记录历史。"""
        self.history.append(self.usage)
        self.usage = TokenUsage()
        return self.history[-1]

    def remaining(self) -> int:
        """剩余可用 Token。"""
        return max(0, self.max_tokens - self.usage.total_tokens)

    def reset(self):
        """重置当前使用量。"""
        self.usage = TokenUsage()

File: atoms/utils/token/budget.c
/**
 * @file budget.c
 * @brief Token预算管理实现
 * @copyright (c) 2026 SPHARX. All Rights Reserved. "From data intelligence emerges."
 */

#include "token.h"
#include <stdlib.h>

struct agentos_token_budget {
    size_t max_tokens;
    size_t used_tokens;
};

agentos_token_budget_t* agentos_token_budget_create(size_t max_tokens) {
    if (max_tokens == 0) return NULL;
    agentos_token_budget_t* budget = (agentos_token_budget_t*)malloc(sizeof(agentos_token_budget_t));
    if (!budget) return NULL;
    budget->max_tokens = max_tokens;
    budget->used_tokens = 0;
    return budget;
}

void agentos_token_budget_destroy(agentos_token_budget_t* budget) {
    free(budget);
}

int agentos_token_budget_add(agentos_token_budget_t* budget, size_t input_tokens, size_t output_tokens) {
    if (!budget) return -1;
    if (budget->used_tokens + input_tokens + output_tokens > budget->max_tokens) {
        return -1;
    }
    budget->used_tokens += input_tokens + output_tokens;
    return 0;
}

size_t agentos_token_budget_remaining(agentos_token_budget_t* budget) {
    if (!budget) return 0;
    return budget->max_tokens - budget->used_tokens;
}

void agentos_token_budget_reset(agentos_token_budget_t* budget) {
    if (budget) budget->used_tokens = 0;
}